src/lxc/lxccontainer.c

```diff
@@ -4840,7 +4840,66 @@ static int do_lxcapi_mount(struct lxc_container *c,
 }
 
 WRAP_API_6(int, lxcapi_mount, const char *, const char *, const char *,
-	   unsigned long, const void *, struct lxc_mount*)
+			unsigned long, const void *, struct lxc_mount*)
+
+static int do_lxcapi_umount2(struct lxc_container *c, const char *target,
+			  unsigned long mountflags, struct lxc_mount *mnt) {
+	pid_t pid, init_pid;
+	int ret = -1;
+
+	if (!c || !c->lxc_conf) {
+		ERROR("Container or configuration is NULL");
+		return -EINVAL;
+	}
+
+	/* Do the fork */
+	pid = fork();
+	if (pid < 0) {
+		SYSERROR("Could not fork");
+		return -1;
+	}
+	
+	if (pid == 0) {
+		init_pid = do_lxcapi_init_pid(c);
+		if (init_pid < 0) {
+			ERROR("Failed to obtain container's init pid");
+			_exit(EXIT_FAILURE);
+		}
+	
+		/* Enter the container namespaces */
+		if (am_guest_unpriv() && !lxc_list_empty(&c->lxc_conf->id_map)) {
+			/* if fully unprivileged */
+			if (!switch_to_ns(init_pid, "user")){
+				ERROR("Failed to enter user namespace");
+				_exit(EXIT_FAILURE);
+			}
+		}
+		if (!switch_to_ns(init_pid, "mnt")) {
+			ERROR("Failed to enter mount namespace");
+			_exit(EXIT_FAILURE);
+		}
+
+		/* Do the mount */
+		ret = umount2(target, mountflags);
+		if (ret < 0) {
+			/* TODO: deal with MNT_EXPIRE? */
```

Comment:
You can remove that comment too. :)

Reply:
Btw, do we actually want to deal with it? :) I mean, the error might be EAGAIN and stuff...

Reply:
Ok, what we should do is:

If MNT_EXPIRE is specified and we get EAGAIN, we should make sure that
`errno` is preserved and that we return `-EAGAIN` as return value to the caller.
This way callers can decide whether they want to call `umount()` again or not.

In general, weshould make sure that on error we always return the negative
`errno` value to the caller.